File: frontend/src/types/scanner.ts
export type ScanStatus =
  | "pending"
  | "running"
  | "completed"
  | "failed"
  | "cancelled"

export type MacdSignal = "bullish" | "bearish" | "neutral"
export type MarketType = "spot" | "swap" | "future"
export type Timeframe =
  | "1m" | "3m" | "5m" | "15m" | "30m" | "1h" | "2h"
  | "4h" | "6h" | "8h" | "12h" | "1d" | "3d" | "1w"
export type BollingerPosition =
  | "oversold"
  | "near_oversold"
  | "neutral"
  | "near_overbought"
  | "overbought"
export type StochasticSignal =
  | "oversold"
  | "overbought"
  | "bullish_cross"
  | "bearish_cross"
  | "neutral"
export type ConfluenceGrade = "F" | "D" | "C" | "B" | "A" | "A+"
export type TrendState = "bullish" | "bearish" | "neutral" | "unavailable"
export type Availability = IndicatorSignalStatus
export type ConfluenceIndicator =
  | "rsi"
  | "trend"
  | "macd"
  | "bollinger"
  | "stochastic"

/** Configuration complète envoyée à la création d'un job scanner. */
export type ScanConfig = {
  exchange_id: string
  market_type: MarketType
  quote: string
  exclude_stable_pairs: boolean
  max_pairs: number | null
  timeframe: Timeframe
  min_ohlcv_bars: number
  max_concurrency: number
  max_retries: number
  retry_delay_seconds: number
  use_rsi: boolean
  rsi_period: number
  rsi_threshold: number
  use_ma: boolean
  use_sma: boolean
  use_ema: boolean
  sma_periods: number[]
  ema_periods: number[]
  ma_timeframes: Timeframe[]
  min_trend_score: number
  use_macd: boolean
  macd_fast_period: number
  macd_slow_period: number
  macd_signal_period: number
  use_bollinger: boolean
  bollinger_period: number
  bollinger_std_dev: number
  use_stochastic: boolean
  stochastic_k_period: number
  stochastic_d_period: number
  stochastic_oversold: number
  stochastic_overbought: number
  use_confluence_score: boolean
  min_confluence_score: number
  confluence_weights: Record<ConfluenceIndicator, number>
  filter_macd_signal: MacdSignal[] | null
  filter_bb_position: BollingerPosition[] | null
  filter_stoch_signal: StochasticSignal[] | null
}

/** Analyse d'une paire renvoyée dans les résultats terminaux du job. */
export type ScanResult = {
  symbol: string
  timeframe: string
  rsi: number | null
  last_close_price: number | null
  last_close_time: string | null
  trend_score: number | null
  trends: Record<string, boolean | null>
  trend_states: Record<string, TrendState>
  trend_net_score: number | null
  moving_averages: Record<string, number>
  macd: number | null
  macd_signal: number | null
  macd_histogram: number | null
  macd_signal_type: MacdSignal | null
  bb_upper: number | null
  bb_middle: number | null
  bb_lower: number | null
  bb_position: BollingerPosition | null
  stoch_k: number | null
  stoch_d: number | null
  stoch_signal: StochasticSignal | null
  confluence_score: number | null
  confluence_grade: ConfluenceGrade | null
  confluence_breakdown: Record<string, number>
  confluence_effective_weights: Record<string, number>
  confluence_details: Record<string, ConfluenceFactorDetail>
  indicator_availability: Record<string, Availability>
  indicator_signals?: IndicatorSignals
}

export type ConfluenceFactorDetail = {
  status: Availability
  raw_value?: unknown
  signal?: unknown
  factor: number | null
  configured_weight: number
  effective_weight: number | null
  contribution: number | null
  reason: string | null
}

/** Snapshot public du job ; `results` n'est présent que sur l'endpoint de résultats. */
export type ScanJob = {
  id: string
  status: ScanStatus
  config: ScanConfig
  progress: {
    processed: number
    total: number
    successful: number
    filtered: number
    errors: number
    percent: number
  }
  result_count: number
  error: string | null
  created_at: string
  started_at: string | null
  completed_at: string | null
  results?: ScanResult[]
}
import type { IndicatorSignals, IndicatorSignalStatus } from "@/types/indicator-signals"

File: backend/app/models/scanner.py
"""Contrats Pydantic publics des jobs, progressions et résultats de scan."""

from __future__ import annotations

from datetime import datetime, timezone
from enum import StrEnum
from typing import Any

from pydantic import BaseModel, Field

from app.core.settings import ScanConfig
from app.domain.indicators import (
    Availability,
    BollingerPosition,
    ConfluenceGrade,
    MacdSignal,
    SignalDirection,
    StochasticSignal,
    TrendState,
)


class IndicatorSignalModel(BaseModel):
    """Miroir Pydantic additif du contrat ``IndicatorSignal`` (voir
    ``app.domain.indicators.types``), pour exposition publique typée dans
    ``ScanResult``/``SignalObservation`` sans dépendre du ``TypedDict``
    interne.
    """

    status: Availability
    direction: SignalDirection
    signal: str | None = None
    state: str | None = None
    strength: float = Field(ge=0.0, le=1.0)
    reason: str | None = None
    raw_value: float | None = None


class ScanStatus(StrEnum):
    """États successifs possibles d'un job de scan."""

    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class ScanProgress(BaseModel):
    """Compteurs de progression d'un scan en cours.

    Pour chaque symbole terminé, exactement un compteur parmi ``successful``,
    ``filtered`` et ``errors`` est incrémenté avec ``processed``.
    """

    processed: int = Field(default=0, description="Nombre de symboles dont l'analyse est terminée.")
    total: int = Field(default=0, description="Nombre total de symboles retenus pour le scan.")
    successful: int = 0
    filtered: int = 0
    errors: int = 0

    @property
    def percent(self) -> float:
        """Calcule la progression sur 100, arrondie à deux décimales."""
        return round((self.processed / self.total) * 100, 2) if self.total else 0.0


class ScanResult(BaseModel):
    """Résultat technique d'une paire ayant franchi tous les filtres actifs.

    Les champs liés à un indicateur désactivé ou non calculable restent à
    ``None``; les collections correspondantes restent vides.
    """

    symbol: str
    timeframe: str
    rsi: float | None = None
    last_close_price: float | None = None
    last_close_time: datetime | None = None
    trend_score: int | None = None
    trends: dict[str, bool | None] = Field(default_factory=dict)
    trend_states: dict[str, TrendState] = Field(default_factory=dict)
    trend_net_score: int | None = None
    moving_averages: dict[str, float] = Field(default_factory=dict)
    macd: float | None = None
    macd_signal: float | None = None
    macd_histogram: float | None = None
    macd_signal_type: MacdSignal | None = None
    bb_upper: float | None = None
    bb_middle: float | None = None
    bb_lower: float | None = None
    bb_position: BollingerPosition | None = None
    stoch_k: float | None = None
    stoch_d: float | None = None
    stoch_signal: StochasticSignal | None = None
    confluence_score: float | None = None
    confluence_grade: ConfluenceGrade | None = None
    confluence_breakdown: dict[str, float] = Field(default_factory=dict)
    confluence_effective_weights: dict[str, float] = Field(default_factory=dict)
    confluence_details: dict[str, dict[str, Any]] = Field(default_factory=dict)
    indicator_availability: dict[str, Availability] = Field(default_factory=dict)
    indicator_signals: dict[str, IndicatorSignalModel] = Field(default_factory=dict)


class ScanJob(BaseModel):
    """État en mémoire d'un scan asynchrone et de ses résultats."""

    id: str = Field(description="Identifiant hexadécimal unique du job.")
    status: ScanStatus = ScanStatus.PENDING
    config: ScanConfig
    progress: ScanProgress = Field(default_factory=ScanProgress)
    results: list[ScanResult] = Field(default_factory=list)
    error: str | None = None
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    started_at: datetime | None = None
    completed_at: datetime | None = None

    def public_payload(self, include_results: bool = False) -> dict[str, Any]:
        """Produit le payload REST/WebSocket stable exposé aux clients.

        Args:
            include_results: Ajoute la liste complète des résultats si vrai.

        Returns:
            Représentation JSON avec ``progress.percent`` et ``result_count``.
        """
        payload = self.model_dump(mode="json", exclude={"results"})
        payload["progress"]["percent"] = self.progress.percent
        payload["result_count"] = len(self.results)
        if include_results:
            payload["results"] = [item.model_dump(mode="json") for item in self.results]
        return payload
